File: TextosJapones/src/main/java/org/jisho/textosJapones/fileparse/RarParse.java
package org.jisho.textosJapones.fileparse;

import com.github.junrar.Archive;
import com.github.junrar.exception.RarException;
import com.github.junrar.rarfile.FileHeader;
import org.jisho.textosJapones.util.Util;

import java.io.*;
import java.util.*;

public class RarParse implements Parse {

	private Archive mArquivo;
	private File mPastaCache;
	private ArrayList<FileHeader> mCabecalhos = new ArrayList<FileHeader>();
	private ArrayList<FileHeader> mLegendas = new ArrayList<FileHeader>();

	@Override
	public void parse(File file) throws IOException {
		try {
			mArquivo = new Archive(file);
		} catch (RarException e) {
			e.printStackTrace();
			throw new IOException("unable to open archive");
		}

		FileHeader cabecalho = mArquivo.nextFileHeader();
		while (cabecalho != null) {
			if (!cabecalho.isDirectory()) {
				String name = getName(cabecalho);
				if (Util.isImage(name)) {
					mCabecalhos.add(cabecalho);
				}
			}

			cabecalho = mArquivo.nextFileHeader();
		}

		Collections.sort(mCabecalhos, new Comparator<FileHeader>() {
			public int compare(FileHeader a, FileHeader b) {
				return Util.getCaminho(getName(a)).compareTo(Util.getCaminho(getName(b)));
			}
		}.thenComparing(new Comparator<FileHeader>() {
			public int compare(FileHeader a, FileHeader b) {
				return Util.getNomeNormalizadoOrdenacao(getName(a)).compareTo(Util.getNomeNormalizadoOrdenacao(getName(b)));
			}
		}));
	}

	@Override
	public int getSize() {
		return mCabecalhos.size();
	}

	@Override
	public InputStream getPagina(int num) throws IOException {
		try {
			FileHeader cabecalho = mCabecalhos.get(num);

			if (mPastaCache != null) {
				String name = getName(cabecalho);
				File cacheFile = new File(mPastaCache, Util.MD5(name));

				if (cacheFile.exists()) {
					return new FileInputStream(cacheFile);
				}

				synchronized (this) {
					if (!cacheFile.exists()) {
						FileOutputStream os = new FileOutputStream(cacheFile);
						try {
							mArquivo.extractFile(cabecalho, os);
						} catch (Exception e) {
							os.close();
							cacheFile.delete();
							throw e;
						}
						os.close();
					}
				}
				return new FileInputStream(cacheFile);
			}
			return mArquivo.getInputStream(cabecalho);
		} catch (RarException e) {
			throw new IOException("unable to parse rar");
		}
	}

	@Override
	public void destroir() throws IOException {
		if (mPastaCache != null) {
			for (File f : mPastaCache.listFiles()) {
				f.delete();
			}
			mPastaCache.delete();
		}
		mArquivo.close();
	}

	@Override
	public String getTipo() {
		return "rar";
	}

	public void setCacheDirectory(File cacheDirectory) {
		mPastaCache = cacheDirectory;
		if (!mPastaCache.exists())
			mPastaCache.mkdirs();
		
		if (mPastaCache.listFiles() != null) {
			for (File f : mPastaCache.listFiles())
				f.delete();
		}
	}

	@Override
	public List<String> getLegenda() {
		List<String> legendas = new ArrayList<String>();
		mLegendas.forEach((it) -> {
			InputStream sub;
			BufferedReader reader;
			try {
				sub = mArquivo.getInputStream(it);
				reader = new BufferedReader(new InputStreamReader(sub, "UTF-8"));
				StringBuilder content = new StringBuilder();

				var line = reader.readLine();
				while (line != null) {
					content.append(line);
					line = reader.readLine();
				}

				legendas.add(content.toString());
			} catch (IOException e) {
				e.printStackTrace();
			}
		});
		return legendas;
	}

	@Override
	public Map<String, Integer> getLegendaNomes() {
		Map<String, Integer> arquivos = new HashMap<String, Integer>();

		for (var i = 0; i < mLegendas.size(); i++) {
			String path = Util.getNome(getName(mLegendas.get(i)));
			if (!path.isEmpty() && !arquivos.containsKey(path))
				arquivos.put(path, i);
		}

		return arquivos;
	}

	private String getName(FileHeader header) {
		return header.getFileName();
	}

	@Override
	public String getPaginaPasta(Integer num) {
		if (mCabecalhos.size() < num)
			return null;
		return getName(mCabecalhos.get(num));
	}

	@Override
	public Map<String, Integer> getPastas() {
		Map<String, Integer> pastas = new HashMap<String, Integer>();

		for (var i = 0; i < mCabecalhos.size(); i++) {
			String path = Util.getPasta(getName(mCabecalhos.get(i)));
			if (!path.isEmpty() && !pastas.containsKey(path))
				pastas.put(path, i);
		}

		return pastas;
	}
}

File: ProcessaTexto/src/main/kotlin/br/com/fenix/processatexto/fileparse/RarParse.kt
package br.com.fenix.processatexto.fileparse

import br.com.fenix.processatexto.util.Utils
import com.github.junrar.Archive
import com.github.junrar.exception.RarException
import com.github.junrar.rarfile.FileHeader
import org.slf4j.LoggerFactory
import java.io.*
import java.nio.charset.StandardCharsets
import java.util.*
import java.util.function.Consumer

class RarParse : Parse {
    private val LOGGER = LoggerFactory.getLogger(RarParse::class.java)

    private lateinit var mArquivo: Archive
    private lateinit var mPastaCache: File
    private val mCabecalhos = ArrayList<FileHeader>()
    private val mLegendas = ArrayList<FileHeader>()

    @Throws(IOException::class)
    override fun parse(file: File) {
        mArquivo = try {
            Archive(file)
        } catch (e: RarException) {
            LOGGER.error(e.message, e)
            throw IOException("unable to open archive")
        }
        var cabecalho = mArquivo.nextFileHeader()
        while (cabecalho != null) {
            if (!cabecalho.isDirectory) {
                val name = getName(cabecalho)
                if (Utils.isImage(name)) {
                    mCabecalhos.add(cabecalho)
                }
            }
            cabecalho = mArquivo.nextFileHeader()
        }
        Collections.sort(mCabecalhos, Comparator<FileHeader> { a, b -> Utils.getCaminho(getName(a)).compareTo(Utils.getCaminho(getName(b))) }
            .thenComparing { a, b -> Utils.getNomeNormalizadoOrdenacao(getName(a)).compareTo(Utils.getNomeNormalizadoOrdenacao(getName(b))) })
    }

    override fun getSize(): Int = mCabecalhos.size

    @Throws(IOException::class)
    override fun getPagina(num: Int): InputStream {
        return try {
            val cabecalho = mCabecalhos[num]
            if (this::mPastaCache.isInitialized) {
                val name = getName(cabecalho)
                val cacheFile = File(mPastaCache, Utils.MD5(name))

                if (cacheFile.exists())
                    return FileInputStream(cacheFile)

                synchronized(this) {
                    if (!cacheFile.exists()) {
                        val os = FileOutputStream(cacheFile)
                        try {
                            mArquivo.extractFile(cabecalho, os)
                        } catch (e: Exception) {
                            cacheFile.delete()
                            throw e
                        } finally {
                            os.close()
                        }
                    }
                }
                FileInputStream(cacheFile)
            } else
                mArquivo.getInputStream(cabecalho)
        } catch (e: RarException) {
            throw IOException("unable to parse rar")
        }
    }

    @Throws(IOException::class)
    override fun destroir() {
        if (this::mPastaCache.isInitialized) {
            for (f in mPastaCache.listFiles()) {
                f.delete()
            }
            mPastaCache.delete()
        }
        mArquivo.close()
    }

    override fun getTipo(): String = "rar"

    fun setCacheDirectory(cacheDirectory: File) {
        mPastaCache = cacheDirectory
        if (!mPastaCache.exists()) mPastaCache.mkdirs()
        if (mPastaCache.listFiles() != null) {
            for (f in mPastaCache.listFiles()) f.delete()
        }
    }

    override fun getLegenda(): List<String> {
        val legendas: MutableList<String> = ArrayList()
        mLegendas.forEach(Consumer { it: FileHeader ->
            val sub: InputStream
            val reader: BufferedReader
            try {
                sub = mArquivo.getInputStream(it)
                reader = BufferedReader(InputStreamReader(sub, StandardCharsets.UTF_8))
                val content = StringBuilder()
                var line = reader.readLine()
                while (line != null) {
                    content.append(line)
                    line = reader.readLine()
                }
                legendas.add(content.toString())
            } catch (e: IOException) {
                LOGGER.error(e.message, e)
            }
        })
        return legendas
    }

    override fun getLegendaNomes(): Map<String, Int> {
        val arquivos: MutableMap<String, Int> = HashMap()
        for (i in mLegendas.indices) {
            val path: String = Utils.getNome(getName(mLegendas[i]))
            if (path.isNotEmpty() && !arquivos.containsKey(path))
                arquivos[path] = i
        }
        return arquivos
    }

    private fun getName(header: FileHeader): String = header.fileName

    override fun getPaginaPasta(num: Int): String {
        return if (mCabecalhos.size < num) "" else getName(mCabecalhos[num])
    }

    override fun getPastas(): Map<String, Int> {
        val pastas: MutableMap<String, Int> = HashMap()
        for (i in mCabecalhos.indices) {
            val path: String = Utils.getPasta(getName(mCabecalhos[i]))
            if (path.isNotEmpty() && !pastas.containsKey(path))
                pastas[path] = i
        }
        return pastas
    }
}